core/services/relay/evm/chain_reader.go

```diff
@@ -104,59 +106,71 @@ func (cr *chainReader) getLatestValueFromContract(ctx context.Context, bc common
 		return err
 	}
 
-	return cr.codec.Decode(ctx, output, returnVal, wrapItemType(method, false))
+	return cr.codec.Decode(ctx, output, returnVal, wrapItemType(bc.Name, method, false))
 }
 
 func (cr *chainReader) Start(_ context.Context) error {
 	return cr.StartOnce("ChainReader", func() error {
-		for name, eventId := range cr.events {
-			if err := cr.lp.RegisterFilter(logpoller.Filter{
-				Name:      name,
-				EventSigs: evmtypes.HashArray{eventId},
-				Addresses: evmtypes.AddressArray{cr.contractID},
-			}); err != nil {
-				return fmt.Errorf("%w: %w", commontypes.ErrInternal, err)
+		for contractName, contractEvents := range cr.events {
+			bindings := cr.bindings[contractName]
+			for name, eventId := range contractEvents {
+				if err := cr.lp.RegisterFilter(logpoller.Filter{
+					Name:      name,
+					EventSigs: evmtypes.HashArray{eventId},
+					Addresses: evmtypes.AddressArray{bindings[name]},
+				}); err != nil {
+					return fmt.Errorf("%w: %w", commontypes.ErrInternal, err)
+				}
 			}
 		}
 		return nil
 	})
 }
 func (cr *chainReader) Close() error {
-	return nil
+	return cr.StopOnce("ChainReader", func() error {
+		for _, contractEvents := range cr.events {
+			for name := range contractEvents {
+				if err := cr.lp.UnregisterFilter(name); err != nil {
+					return fmt.Errorf("%w: %w", commontypes.ErrInternal, err)
+				}
+			}
+		}
+		return nil
+	})
 }
 
 func (cr *chainReader) Ready() error { return nil }
 func (cr *chainReader) HealthReport() map[string]error {
 	return map[string]error{cr.Name(): nil}
 }
 
-func (cr *chainReader) CreateType(itemType string, forEncoding bool) (any, error) {
-	return cr.codec.CreateType(wrapItemType(itemType, forEncoding), forEncoding)
+func (cr *chainReader) CreateContractType(contractName, itemType string, forEncoding bool) (any, error) {
+	return cr.codec.CreateType(wrapItemType(contractName, itemType, forEncoding), forEncoding)
 }
 
-func addEventTypes(name string, contractABI abi.ABI, chainReaderDefinition types.ChainReaderDefinition, parsed *parsedTypes) (common.Hash, error) {
+func addEventTypes(contractName, typeName string, contractABI abi.ABI, chainReaderDefinition types.ChainReaderDefinition, parsed *parsedTypes) (common.Hash, error) {
 	event, methodExists := contractABI.Events[chainReaderDefinition.ChainSpecificName]
 	if !methodExists {
 		return common.Hash{}, fmt.Errorf("method: %s doesn't exist", chainReaderDefinition.ChainSpecificName)
 	}
 
-	return event.ID, addDecoderDef(name, event.Inputs, parsed, chainReaderDefinition)
+	return event.ID, addDecoderDef(contractName, typeName, event.Inputs, parsed, chainReaderDefinition)
 }
 
-func addMethods(name string, abi abi.ABI, chainReaderDefinition types.ChainReaderDefinition, parsed *parsedTypes) error {
+func addMethods(contractName, typeName string, abi abi.ABI, chainReaderDefinition types.ChainReaderDefinition, parsed *parsedTypes) error {
```

Comment:
Not really related to this PR, but just realised that these should be renamed to `addEventType` and add `addMethodType`. And maybe add comment that explains that by add we mean add to codec.